File: System-Design/Distributed-Consensus/Raft/raft_node.py
import random

class NodeState:
    FOLLOWER = "FOLLOWER"
    CANDIDATE = "CANDIDATE"
    LEADER = "LEADER"

class RaftNode:
    def __init__(self, node_id: int):
        self.node_id = node_id
        self.state = NodeState.FOLLOWER
        self.current_term = 0
        self.voted_for = None
        self.leader_id = None
        
        # Raft uses randomized timeouts (150ms - 300ms) to prevent split votes
        self.election_timeout = random.randint(150, 300)
        self.timer = 0

    def tick(self, delta_ms: int):
        if self.state == NodeState.LEADER:
            # Leaders don't timeout, they send heartbeats
            return

        self.timer += delta_ms
        if self.timer >= self.election_timeout:
            self._start_election()

    def _start_election(self):
        self.state = NodeState.CANDIDATE
        self.current_term += 1
        self.voted_for = self.node_id
        self.timer = 0
        self.election_timeout = random.randint(150, 300)
        # In a full network, we would broadcast RequestVote RPCs here

    def receive_heartbeat(self, leader_id: int, leader_term: int) -> bool:
        # If we see a term greater than or equal to ours, we step down and follow
        if leader_term >= self.current_term:
            self.state = NodeState.FOLLOWER
            self.current_term = leader_term
            self.leader_id = leader_id
            self.voted_for = None
            self.timer = 0 # Reset our election timer!
            return True
        return False # Reject obsolete leaders

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    node = RaftNode(node_id=1)
    
    assert node.state == NodeState.FOLLOWER, "Node should start as Follower"
    
    # Simulate time passing to trigger a timeout
    node.tick(400)
    
    assert node.state == NodeState.CANDIDATE, "Node failed to transition to Candidate on timeout"
    assert node.voted_for == 1, "Candidate must vote for itself"
    
    # Simulate a network event: A different node won the election and sent a heartbeat with a higher term
    accepted = node.receive_heartbeat(leader_id=2, leader_term=node.current_term + 1)
    
    assert accepted == True, "Node rejected a valid higher-term leader"
    assert node.state == NodeState.FOLLOWER, "Node failed to step down gracefully"
    assert node.leader_id == 2, "Node did not recognize the new leader"
    assert node.timer == 0, "Node failed to reset its election timer after heartbeat"
    
    print("Python Raft Consensus State Machine Test Passed!")

File: System-Design/Distributed-Consensus/Raft/raftNode.ts
enum NodeState {
    FOLLOWER = "FOLLOWER",
    CANDIDATE = "CANDIDATE",
    LEADER = "LEADER"
}

export class RaftNode {
    public nodeId: number;
    public state: NodeState;
    public currentTerm: number;
    public votedFor: number | null;
    public leaderId: number | null;
    
    private electionTimeout: number;
    public timer: number;

    constructor(nodeId: number) {
        this.nodeId = nodeId;
        this.state = NodeState.FOLLOWER;
        this.currentTerm = 0;
        this.votedFor = null;
        this.leaderId = null;
        
        // Random timeout between 150ms and 300ms
        this.electionTimeout = Math.floor(Math.random() * (300 - 150 + 1) + 150);
        this.timer = 0;
    }

    public tick(deltaMs: number): void {
        if (this.state === NodeState.LEADER) return;

        this.timer += deltaMs;
        if (this.timer >= this.electionTimeout) {
            this.startElection();
        }
    }

    private startElection(): void {
        this.state = NodeState.CANDIDATE;
        this.currentTerm++;
        this.votedFor = this.nodeId;
        this.timer = 0;
        this.electionTimeout = Math.floor(Math.random() * (300 - 150 + 1) + 150);
    }

    public receiveHeartbeat(leaderId: number, leaderTerm: number): boolean {
        if (leaderTerm >= this.currentTerm) {
            this.state = NodeState.FOLLOWER;
            this.currentTerm = leaderTerm;
            this.leaderId = leaderId;
            this.votedFor = null;
            this.timer = 0; // Reset timer
            return true;
        }
        return false;
    }
}

// --- CI/CD Automated Test ---
const node = new RaftNode(1);

if (node.state !== NodeState.FOLLOWER) process.exit(1);

node.tick(400);

if (node.state !== NodeState.CANDIDATE || node.votedFor !== 1) process.exit(1);

const accepted = node.receiveHeartbeat(2, node.currentTerm + 1);

if (accepted && node.state === NodeState.FOLLOWER && node.leaderId === 2 && node.timer === 0) {
    console.log("TypeScript Raft Consensus State Machine Test Passed!");
} else {
    process.exit(1);
}